### PR TITLE
Bug 1195539 - Filter out non-l10n elements from translateMutations and remove condition race

### PR DESCRIPTION
**Pull request by zbraniecki**
https://bugzilla.mozilla.org/show_bug.cgi?id=1195539

### CODE REVIEW

File: src/bindings/html/dom.js [hunk 1]
```diff
@@ -42,15 +42,14 @@ export function getAttributes(element) {
 }
 
 function getTranslatables(element) {
-  const nodes = [];
+  const nodes = [...element.querySelectorAll('*[data-l10n-id]')];
```
Comment (stasm) [inline src/bindings/html/dom.js]: I really like this way of transforming a `NodeList` into an array! It does seem a bit slower than the traditional methods (cf. http://jsperf.com/qsa-slice-vs-spread) but I'm all for keeping it here.

File: src/bindings/html/dom.js [hunk 2]
```diff
@@ -42,15 +42,14 @@ export function getAttributes(element) {
...
 
-  if (typeof element.hasAttribute === 'function' &&
+  if (element.nodeType !== element.DOCUMENT_FRAGMENT_NODE &&
```
Comment (stasm) [inline src/bindings/html/dom.js]: Why this change?  The original served the same purpose and was more generic and possibly future-proof.

File: src/bindings/html/dom.js [hunk 3]
```diff
@@ -77,19 +76,31 @@ export function translateMutations(view, langs, mutations) {
 
   const elements = [];
 
-  targets.forEach(target => target.childElementCount ?
-      elements.push(...getTranslatables(target)) : elements.push(target));
+  targets.forEach(target => {
+    if (target.childElementCount) {
+      elements.push(...getTranslatables(target));
+    } else {
+      if (target.hasAttribute('data-l10n-id')) {
```
Comment (stasm) [inline src/bindings/html/dom.js]: You can collapse this into a `else if`.  I think I preferred the ternary version but that's just style.

Comment (zbraniecki) [inline src/bindings/html/dom.js]: the ternary style would require me to have an if in the else structure. I feel like we're paying with code readability for dense coding at some point. if/else block may be longer but I feel like it's more readable. Would you agree?

File: src/bindings/html/dom.js (hunk 2 above)
Comment (zbraniecki) [inline src/bindings/html/dom.js]: I think I thought that it's smarter... but I'm not sure anymore. Will revert.

File: src/bindings/html/dom.js (hunk 3 above)
Comment (stasm) [inline src/bindings/html/dom.js]: It all boils down to personal preference, and I'm okay with your approach.

I think the following is also quite readable and I like how it's a single expression.

```javascript
targets.forEach(target => target.childElementCount ?
  elements.push(...getTranslatables(target)) :
  target.hasAttribute('data-l10n-id') ?
    elements.push(target)) : undefined
```

Again, this is just personal preference.  r+ on your version! :)

File: src/bindings/html/dom.js [hunk 4]
```diff
@@ -77,19 +76,31 @@ export function translateMutations(view, langs, mutations) {
...
+  if (elements.length) {
+    translateElements(view, langs, elements);
+  }
+}
 
-  Promise.all(
-    elements.map(elem => getElementTranslation(view, langs, elem))).then(
-      translations => applyTranslations(view, elements, translations));
+function translateElements(view, langs, elements) {
+  return view.ready.then(() => {
```
Comment (stasm) [inline src/bindings/html/dom.js]: I'm still thinking about this :)

File: src/bindings/html/dom.js (hunk 1 above)
Comment (stasm) [inline src/bindings/html/dom.js]: Also, and it's just a nit:  I think we can remove the `*` from the qSA.

File: src/bindings/html/dom.js (hunk 4 above)
Comment (stasm) [inline src/bindings/html/dom.js]: I like that we're not repeating ourselves here and that you abstracted `translateElements` into its own function.  Can you move it closer to where `translateElement` is defined?

However, I think that the `view.ctx.fetch` guard should be set in view.js, not here.  In dom.js, we make certain assumptions and one of them is that the view takes care of guaranteeing that the first language has been loaded.

(We can discuss if this assumption is valid.  For now, I suggest we keep it and keep things consistent.)

Take a look at `translateFragment` in dom.js and in view.js.  The view calls `ctx.fetch` before it calls dom's `translateFragment`.  I think we should do the same for mutations:  in view's `onMutations` we should add `ctx.fetch`, just like it's done a few lines above in `translateFragment`.

What do you think?

Comment (zbraniecki) [inline src/bindings/html/dom.js]: Great! Updated the patch.

File: src/bindings/html/dom.js [hunk 5]
```diff
@@ -42,15 +42,14 @@ export function getAttributes(element) {
 }
 
 function getTranslatables(element) {
-  const nodes = [];
+  const nodes = [...element.querySelectorAll('[data-l10n-id]')];
 
-  if (typeof element.hasAttribute === 'function' &&
+  if (element.nodeType !== element.DOCUMENT_FRAGMENT_NODE &&
```
Comment (stasm) [inline src/bindings/html/dom.js]: Forgot to revert this?

File: src/bindings/html/dom.js (hunk 3 above)
Comment (stasm) [inline src/bindings/html/dom.js]: Can you still collapse the `else { if` into an `else if`?

File: src/bindings/html/dom.js [hunk 6]
```diff
@@ -77,19 +76,30 @@ export function translateMutations(view, langs, mutations) {
...
+    }
+  });
 
-  Promise.all(
+  if (elements.length) {
+    translateElements(view, langs, elements);
+  }
+}
+
+function translateElements(view, langs, elements) {
```
Comment (stasm) [inline src/bindings/html/dom.js]: And put this closer to `translateElement`, please?

**Pull request merged**